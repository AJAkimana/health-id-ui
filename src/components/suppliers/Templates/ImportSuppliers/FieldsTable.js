import React from 'react';

const FieldsTable = () => (
  <div>
    <div>
      <h3 className="fields-title">Ensure the following fields are present in the .csv file you are uploading</h3>
    </div>
    <table className="fields-table">
      <tbody>
        <tr className="fields-tr">
          <td className="required-field">
            - Name
          </td>
          <td className="required-field">
            - Email
          </td>
          <td className="required-field">
            - Mobile #
          </td>
          <td className="required-field">
            - Tier
          </td>
        </tr>
        <tr>
          <td className="required-field">
            - Address Line 1
          </td>
          <td className="required-field">
            - Address Line 2
          </td>
          <td className="required-field">
            - Country
          </td>
          <td className="required-field">
            - City
          </td>
        </tr>
        <tr>
          <td className="required-field">
            - LGA
          </td>
          <td className="required-field">
            - Payment terms
          </td>
          <td className="required-field">
            - Credit Days
          </td>
          <td className="required-field">
            - Supplier Rating
          </td>
        </tr>
        <tr>
          <td className="required-field">
            - Commentary
          </td>
          <td className="required-field">
            - Supplier Logo
          </td>
        </tr>
      </tbody>
    </table>
  </div>
);

export default FieldsTable;
