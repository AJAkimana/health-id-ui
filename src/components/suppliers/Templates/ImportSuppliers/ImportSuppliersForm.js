import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import {
  Paper, Grid, Typography, Button
} from '@material-ui/core';
import UploadFileComponent from '../../../shared/UploadFileComponent';
import { ImportProductStyles } from '../../../../assets/styles/products/ImportProductStyles';
import SubmitFile from '../../../shared/SubmitFileComponent';
import FieldsTable from './FieldsTable';
import Separator from '../../../shared/Separator';


const ImportSuppliersForm = (props) => {
  const {
    state: {
      file, loading, serverResponse
    },
    onDrop,
    handleFile,
    handleUpload,
    handleDownloadTemplate,
    fileLoaded,
    isSubmitFileFailed,
    handleUploadFailed
  } = props;

  const disableDownload = !!file;

  const ButtonStyle = file ? ImportProductStyles.disabledBtn : ImportProductStyles.templateBtn;

  return (
    <Fragment>
      <Paper
        style={ImportProductStyles.paper}
      >
        <Grid
          container
          spacing={24}
          style={ImportProductStyles.importContainer}
        >
          <Grid
            item
            xs={12}
            style={ImportProductStyles.textSection}
          >
            <Typography
              style={ImportProductStyles.typography}
            >
              Upload Manager
            </Typography>
          </Grid>

          <Grid
            item
            xs={12}
          >
            <FieldsTable />
          </Grid>
          <Grid
            item
            xs={12}
          >
            <Separator />
          </Grid>
          <Grid
            item
            xs={12}
          >
            <div style={ImportProductStyles.templateDiv}>
              <Button
                style={ButtonStyle}
                disabled={disableDownload}
                onClick={handleDownloadTemplate}
              >
                DOWNLOAD SAMPLE TEMPLATE
              </Button>
            </div>
          </Grid>
        </Grid>
      </Paper>
      {!fileLoaded ? (
        <UploadFileComponent
          onDrop={onDrop}
          handleFile={handleFile}
        />
      ) : (
        <SubmitFile
          file={file}
          loading={loading}
          handleUpload={handleUpload}
          serverResponse={serverResponse}
          isSubmitFileFailed={isSubmitFileFailed}
          handleUploadFailed={handleUploadFailed}
        />
      )
      }
    </Fragment>
  );
};

ImportSuppliersForm.propTypes = {
  state: PropTypes.instanceOf(Object).isRequired,
  fileLoaded: PropTypes.bool.isRequired,
  onDrop: PropTypes.func,
  handleFile: PropTypes.func,
  handleUpload: PropTypes.func,
  handleDownloadTemplate: PropTypes.func,
  serverResponse: PropTypes.func,
  isSubmitFileFailed: PropTypes.bool.isRequired,
  handleUploadFailed: PropTypes.func.isRequired,
};

ImportSuppliersForm.defaultProps = {
  onDrop: () => {},
  handleFile: () => {},
  handleUpload: () => {},
  handleDownloadTemplate: () => {},
  serverResponse: () => {}
};


export default ImportSuppliersForm;
