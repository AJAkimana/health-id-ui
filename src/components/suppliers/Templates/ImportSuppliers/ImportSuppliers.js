import React, { Component, Fragment } from 'react';
import axios from 'axios';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import withAuth from '../../../withAuth';
import ImportSuppliersForm from './ImportSuppliersForm';
import BackAction from '../../../shared/BackAction';
import verifyFile from '../../../../utils/products/verifyFile';
import notify from '../../../shared/Toaster';

import { StateContext } from '../../../../providers/stateProvider';

export class ImportSuppliers extends Component {
  state = {
    file: null,
    loading: false,
    serverResponse: '',
    fileLoaded: false,
    isSubmitFileFailed: false,
  };

  componentDidMount() {
    const [, dispatch] = Object.values(this.context);
    dispatch({
      type: 'changeGrid',
      grid: 'grid4'
    });
  }

  onDrop = async (acceptedFiles) => {
    this.setState({ file: acceptedFiles[0] });
    await this.handleUpload();
  }

  handleFile = (e) => {
    const { files } = e.target;
    const maxFileSize = 10000000; // bytes
    const acceptedFileType = 'text/csv';
    if (files && files.length > 0) {
      const verified = verifyFile(files, maxFileSize, acceptedFileType);
      if (verified) {
        this.setState({ file: files[0], fileLoaded: true, isSubmitFileFailed: false });
      }
    }
  }

  handleUploadFailed = async () => {
    this.setState({ fileLoaded: false, isSubmitFileFailed: true, serverResponse: '' });
  }

  handleDownloadTemplate = async () => {
    const csvUrl = `${process.env.APP_LINK}`;
    const token = localStorage.getItem('rest_token');
    /* istanbul ignore next */
    axios.get(`${csvUrl}sample_csv_file/suppliers`, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Token ${token}`,
      },
    })
      .then((res) => {
        const filename = 'supplier_template.csv';
        const blob = new Blob([res.data], { type: 'text/csv' });
        const link = document.createElement('a');
        link.href = window.URL.createObjectURL(blob);
        link.download = filename;
        document.body.appendChild(link);
        window.location.replace(link);
        link.click();
        link.parentNode.removeChild(link);
      })
      .catch((err) => {
        notify('There was a problem downloading this file.', err);
      });
  }

  handleUpload = async () => {
    this.setState({ loading: true });

    const { file } = this.state;
    const { history } = this.props;
    const formdata = new FormData();

    formdata.append('file', file);

    const url = `${process.env.APP_LINK}`;
    const token = localStorage.getItem('rest_token');

    try {
      const res = await axios.post(`${url}csv/suppliers`, formdata, {
        headers: {
          'Content-Type': 'multipart/form-data',
          Authorization: `Token ${token}`,
        }
      });
      const { noOfSuppliersAdded } = res.data;

      this.setState({
        loading: false,
        serverResponse: `${noOfSuppliersAdded} suppliers have been successfully uploaded for approval`,
      });
      setTimeout(() => history.push('/suppliers'), 1200);
    } catch (err) {
      this.setState({
        loading: false,
        isSubmitFileFailed: true,
        serverResponse: err.status === 500 ? 'Something went wrong' : err.response.data.error,
      });
    }
  }

  static contextType = StateContext;

  render() {
    const { fileLoaded, isSubmitFileFailed } = this.state;
    return (
      <Fragment>
        <BackAction
          header="Back"
          link="/suppliers"
        />
        <ImportSuppliersForm
          state={this.state}
          fileLoaded={fileLoaded}
          isSubmitFileFailed={isSubmitFileFailed}
          onDrop={this.onDrop}
          handleFile={this.handleFile}
          handleUpload={this.handleUpload}
          handleUploadFailed={this.handleUploadFailed}
          handleDownloadTemplate={this.handleDownloadTemplate}
        />
      </Fragment>
    );
  }
}

ImportSuppliers.propTypes = {
  history: PropTypes.shape({
    lenght: PropTypes.number,
    push: PropTypes.func
  })
};

ImportSuppliers.defaultProps = {
  history: PropTypes.shape({
    lenght: PropTypes.number,
    push: PropTypes.func
  })
};

export default withAuth(withRouter(ImportSuppliers));
