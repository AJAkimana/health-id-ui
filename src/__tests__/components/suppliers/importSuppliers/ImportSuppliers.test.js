import React from 'react';
import { shallow } from 'enzyme';
import * as moxios from 'moxios';
import toJson from 'enzyme-to-json';
import PropTypes from 'prop-types';
import { ImportSuppliers } from '../../../../components/suppliers/Templates/ImportSuppliers/ImportSuppliers';


describe('Render ImportSuppliers Component', () => {
  const props = {
    session: {
      me: {
        mobileNumber: '2348903345',
        email: 'user@gmail.com',
        username: 'user',
        role: { name: 'Master Admin' },
      }
    }
  };

  const acceptedFiles = [{
    name: 'supplier', size: 10000, type: 'text/csv', lastModified: ''
  }];

  const expectedFile = {
    name: 'supplier', size: 10000, type: 'text/csv', lastModified: ''
  };

  global.setTimeout = jest.fn();

  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  ImportSuppliers.contextTypes = [
    PropTypes.string,
    PropTypes.func
  ];

  const context = ['kitty', jest.fn()]

  const wrapper = shallow(<ImportSuppliers {...props} />, { context });
  it('renders without crashing', () => {
    expect(wrapper.find('BackAction').length).toBe(1);
    expect(wrapper.find('ImportSuppliersForm').length).toBe(1);
  });

  it('should render the ImportSuppliers component properly', () => {
    const setupImportSuppliers = () => {
      const wrapper = shallow(<ImportSuppliers {...props} />, { context })
      return wrapper
    }

    let wrapper;
    wrapper = setupImportSuppliers();
    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('calls handleFile', () => {
    const newProps = {
      state: { file: "" },
      ...props
    };

    const e = {
      target: {
        files: acceptedFiles,
      }
    };
    const wrapperWithFileProp = shallow(<ImportSuppliers {...newProps} />, { context });
    wrapperWithFileProp.instance().handleFile(e);
    expect(wrapperWithFileProp.state().file).toEqual(expectedFile);
  });

  it('calls onDrop', () => {
    wrapper.instance().onDrop(acceptedFiles);
    expect(wrapper.state().file).toEqual(expectedFile);
  });

  it('calls handleDownloadTemplate', async () => {
    const url = `${process.env.APP_LINK}/sample_csv_file/supplier`;
    wrapper.instance().handleDownloadTemplate();
    moxios.stubRequest(url, {
      status: 200,
      res: {
        data: [
          'Name, Email,	Mobile #,	Supplier Rating, Address Line 1,	Address Line 2,	LGA,	City,	Tier,	Supplier Logo,	Commentary,	Payment terms,	Credit days'
        ]
      }
    });
  });

  it('calls handleUpload', async () => {
    const url = `${process.env.APP_LINK}/csv/suppliers`;
    wrapper.instance().handleUpload();
    wrapper.instance().handleUploadFailed().then(() => { });
    expect(wrapper.state().loading).toEqual(true);
    moxios.stubRequest(url, {
      status: 200,
      res: {
        data: {
          noOfProductsAdded: 20
        }
      }
    });
  });
});
