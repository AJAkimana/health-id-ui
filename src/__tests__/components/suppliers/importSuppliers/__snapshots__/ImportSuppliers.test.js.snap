// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ImportSuppliers Component should render the ImportSuppliers component properly 1`] = `
<Fragment>
  <BackAction
    header="Back"
    link="/suppliers"
  />
  <ImportSuppliersForm
    fileLoaded={false}
    handleDownloadTemplate={[Function]}
    handleFile={[Function]}
    handleUpload={[Function]}
    handleUploadFailed={[Function]}
    isSubmitFileFailed={false}
    onDrop={[Function]}
    serverResponse={[Function]}
    state={
      Object {
        "file": null,
        "fileLoaded": false,
        "isSubmitFileFailed": false,
        "loading": false,
        "serverResponse": "",
      }
    }
  />
</Fragment>
`;
